### PR TITLE
Log the element type and UUID for FHIR elements filtered out by Condition Filters

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR updates the logging of filter results on submitted reports. When filters are run on specific parts of the report's Bundle, the type and UUID for the filtered element is included on the log.

Note - This PR has attached a CURL call that I used on Postman to test. Feel free to use it or make your own test.

Test Steps:
1. In organizations.yml, update the `Ignore FULL_ELR` receiver to have a conditionFilter that will cause at least one `OBX` segment to be filtered out. e.g. `conditionFilter: ["%obsPerformedCodes.intersect('123-4').exists()"]`
2. Run ReportStream, then run `./gradlew reloadSettings` to update the settings.
3. Submit a report with data that would go to the `Ignore FULL_ELR_FHIR` Receiver.
4. Using the History API, verify that at least one row is inside `filteredReportRows` for `Ignore FULL_ELR_FHIR` in the `destinations` list.
5. Verify that the output on the Report Rows matches the format: `For ignore.FULL_ELR_FHIR, filter [FILTERDETAILS] filtered out item [ITEMID] at Observation/[UUID]"`, with relevant values replacing the bracketed contents in the example.

## Changes
- Update `FHIRRouter` to include the part of the item that's been filtered out.

## Linked Issues
- #8689  

[local call curl.txt](https://github.com/CDCgov/prime-reportstream/files/11534718/local.call.curl.txt)

### CODE REVIEW

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:578]: If we instead just check for `focusResource != bundle` here, we can always pass `focusResource` through; we wouldn't need `focusResource` to be nullable or have a default value here, and we wouldn't need the if/else at L436. Works either way, just an idea that I think simplifies this a little bit.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:567]: If you take my suggestion my other comment, then I'd suggest a slight rewording here.
```suggestion
     * @param focusResource Starting point for the evaluation, logged if different than [bundle]
```

**Review by jimmyfagan — COMMENTED** [review]
Just a small optional suggestion, otherwise looks good! Still need to test locally, will approve after that.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:579]: @JessicaWNava the result of this is a filter log message like this: `For ignore.FULL_ELR, filter [%obsPerformedCodes.intersect('123-4').exists()][] filtered out item MT_IGCAA_ORU_AAPHELR.1.6214638 at Observation/1685109567290401000.51a5d5cd-f839-4161-af20-315a2e526a1d`. Is that the kind of ID you were referring to in the description on [the ticket](https://github.com/CDCgov/prime-reportstream/issues/8689)? I'm a little wary of using that, since it only really carries meaning in the context of our internal FHIR bundle (or if the report was FHIR to begin with). Still helps if you pull down the FHIR bundle from object storage to look up that observation, but I'm wondering if instead using the LOINC code would be more useful. Any thoughts?

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Yeah, it's a lot more readable with the changes

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:575]: I'm fine waiting to merge until we're 100% sure of what will be the most useful log message content

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jack-h-wang) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:436]: Existing unit tests for `logFilterResults` are no longer passing with this pattern change.

**Review by jack-h-wang — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:575]: How difficult would it be to use the observation id and loinc code? Honestly, this was a ticket Carlos had me write up. I am not sure what he had in mind, so it's really kind of a "use our best judgement" sort of thing. The goal is for the sender to be able to tell which one didn't make it through

Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Using the LOINC code is not too bad when we know it's an Observation. Made updates to use it. Additionally, it shows the codesystem being used. e.g. `"For ignore.FULL_ELR, filter [%obsPerformedCodes.intersect('123-4').exists()][] filtered out item MT_IGCAA_ORU_AAPHELR.1.6214638 with http://loinc.org code:19000-9",`

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:581]: This works fine, although it takes out the idBase piece of it. I think the latest proposal was to use both the code and the id, would that still make sense?

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:577]: Typo maybe?
```suggestion
            // for Observation-type elements, we use the code property when available
```

**Review by jimmyfagan — COMMENTED** [review]
A few hopefully quick/minor things, but this looks promising! Let me know when it's ready for another look!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:581]: I think it makes sense either way actually, the id is mostly only useful for us to debug using the internal fhir, so I will defer to whatever you decide.

**Review by jimmyfagan — APPROVED** [review]
Since my comments are all either optional or limited to comments, I'll mark this approved, if anything warrants another look, let me know, otherwise, you can let Arnej know this is ready for his pass. Thanks!

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt
Comment (arnejduranovic) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:951]: This isn't actually testing your new functionality? We need unit tests that test that different paths within `logFilterResults`

**Review by arnejduranovic — CHANGES_REQUESTED** [review]
Please update ticket with migration plan for staging and production. You are changing the settings for the Ignore org which can potentially impact smoke tests.

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt
Comment (TurkeyFried) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:951]: Yeah that's true. I just added a new test to cover the new logging functionality

**Review by arnejduranovic — APPROVED** [review]

**Review by jack-h-wang — APPROVED** [review]

**Pull request merged**
